File: py-110/small_problems/medium_1/rotation_3.py
"""
Rotation (Part 3)
Take the number 735291 and rotate it by one digit to the left, getting 352917. Next, keep the first digit fixed in place and rotate the remaining digits to get 329175. Keep the first two digits fixed in place and rotate again to get 321759. Keep the first three digits fixed in place and rotate again to get 321597. Finally, keep the first four digits fixed in place and rotate the final two digits to get 321579. The resulting number is called the maximum rotation of the original number.

Write a function that takes an integer as an argument and returns the maximum rotation of that integer. You can (and probably should) use the rotate_rightmost_digits function from the previous exercise.

ExamplesCopy Code

P:
Input: Integer
Output: Integer
Rules:
    - Explicit
        - Maximum rotation means to iterate through each "place" of the digit, and remove and add the element at that place to the very end. Start with the first digit and rotate, then rotate the second digit of the resulting integer, all the way to the second to last digit of the integer
    - Implicit
        - Assume that we'll have a valid integer argument
        - If input is one digit, return the input

E: Confirmed

D: Intermediary string to work with the integer , as well as an intermediary list to work with the individual digits. We may want to use a range to iterate through the indices of one of the intermediary objects we use

A:
    - Coerce the input into a string, assign to variable 'input_str'
    - For an index that ranges from length of the 'input_str' to 1 (exclusive), incrementing downward as we go
        - Use the rotate_rightmost_digits function, passing in the current value of the input argument as the first argument and the current index as the second argument
        - Reassign the input argument to the result of the prior step
    - Return the latest value of the input argument

"""

def rotate_rightmost_digits(num, places):
    input_str = str(num)
    digits = list(input_str)

    idx = -places
    new_last_element = digits.pop(idx)
    digits.append(new_last_element)

    return int(''.join(digits))

def max_rotation(number):
    input_str = str(number)
    for idx in range(len(input_str), 1, -1):
        return_value = rotate_rightmost_digits(number, idx)
        number = return_value
    return number


print(max_rotation(735291) == 321579)          # True
print(max_rotation(3) == 3)                    # True
print(max_rotation(35) == 53)                  # True
print(max_rotation(8703529146) == 7321609845)  # True

# Note that the final sequence here is `015`. The leading
# zero gets dropped, though, since we're working with
# an integer.
print(max_rotation(105) == 15)                 # True



"""
Rotation (Part 1): Start time 5:19 to 5:18
Rotation (Part 2): start time 5:36 to 5:45
Rotation (Part 3): start time 5:56 to 6:10

"""

File: ls-215/exercises/medium_1/rotation_3.js
/*
Take the number 735291 and rotate it by one digit to the left, getting 352917. Next, keep the first digit fixed in place and rotate the remaining digits to get 329175. Keep the first two digits fixed in place and rotate again to get 321759. Keep the first three digits fixed in place and rotate again to get 321597. Finally, keep the first four digits fixed in place and rotate the final two digits to get 321579. The resulting number is called the maximum rotation of the original number.

Write a function that takes an integer as an argument and returns the maximum rotation of that integer. You can (and probably should) use the rotateRightmostDigits function from the previous exercise.
*/

function rotateDigits(digits) {
  let subArray = digits.slice(1);
  subArray.push(digits[0]);
  return subArray;
}

function rotateRightmostDigits(number, n) {
  n = n > String(number).length ? String(number).length : n;

  let digits = String(number).split('').map(Number);
  let toRotate = digits.slice(-n);
  let rotated = rotateDigits(toRotate);

  let newDigits = digits.slice(0, -n).concat(rotated);
  return Number(newDigits.join(''));
}

/*
My questions:
- Does fully rotated mean what I said it does below?
- (I'll assume the same assumptions of the prior 2 parts of this question hold to keep from re-asking the same things)


P:
Input: number
Output: Fully rotated number
Rules:
- Fully rotated means, for a number with `n` digits:
  - Rotate the last `n` digits of the number
  - Rotate the last `n - 1` digits of the resulting number
  - ...
  - Rotate the last 2 digits of the resulting number
  - No need to rotate the last 1 digit, that doesn't do anything


Data structure:
- (We do a lot of what we need already in rotateRightmostDigits)
- Convert input into string to be able to work with length



Algorithm:
- Convert input into string, get its length
- Set `result` to the value of the input
- For an `n` from the string length to 1 (exclusive):
  - Get the new rotated number
    - Use rotateRightmostDigits, pass in `result` and `n`
  - Reassign `result` to the newly rotated number
- Return `result`

*/

function maxRotation(number) {
  const strLen = String(number).length;
  let result = number;
  for (let n = strLen; n > 1; n -= 1) {
    result = rotateRightmostDigits(result, n);
  }
  return result;
}

console.log(maxRotation(735291));          // 321579
console.log(maxRotation(3));               // 3
console.log(maxRotation(35));              // 53
console.log(maxRotation(105));             // 15 -- the leading zero gets dropped
console.log(maxRotation(8703529146));      // 7321609845


/* EDIT: COULD HAVE BEEN MORE PRECISE HERE WITH LEADING ZEROS
For example, what should happen if we call maxRotation(1005)? Our current code returns 51, is
that what we want? That's what we need to ask questions on.
*/